### PR TITLE
Simplify class stucture of hit generators

### PR DESCRIPTION
**Pull request by makortel**
Current inheritance chain of the hit generators is somewhat messy, and in some parts does not serve the current usage of the classes. This PR removes unnecessary inheritances and does some further cleanup allowed by the removal.

Essentially, the inheritance from HitPairGenerator is removed from
- HitPairGeneratorFromLayerPair
- CombinedHitQuadrupletGeneratorForPhotonConversion
- HitQuadrupletGeneratorFromLayerPairForPhotonConversion
- CosmicHitPairGenerator
- CosmicHitPairGeneratorFromLayerPair,

the inheritance from HitTripletGenerator from
- HitTripletGeneratorFromPairAndLayers
- CosmicHitTripletGenerator
- CosmicHitTripletGeneratorFromLayerTriplet,

and the inheritance from MultiHitGenerator from
- MultiHitGeneratorFromPairAndLayers.

(Most of the developments were actually done already more than a year ago, but for some reason I didn't push them for integration and they got buried under other stuff)

Tested in 750pre5, no changes expected in results.

@rovere @VinInn

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test
[not sure where the labels are, could be the bot didn't look at it yet, unless it's stuck]

**Comment by VinInn** [comment]
Lovely!

File: RecoTracker/TkSeedGenerator/plugins/MultiHitGeneratorFromChi2.cc [hunk 1]
```diff
@@ -334,7 +322,7 @@ void MultiHitGeneratorFromChi2::hitSets(const TrackingRegion& region,
     for(int il = 0; (il < size) & (!usePair); il++) {
 
       if (debugPair) 
-	cout << "cosider layer: " << theLayers[il].name() << " for this pair. Location: " << theLayers[il].detLayer()->location() << endl;
+	cout << "cosider layer: " << thirdLayers[il].name() << " for this pair. Location: " << thirdLayers[il].detLayer()->location() << endl;
```
Comment (cvuosalo) [inline RecoTracker/TkSeedGenerator/plugins/MultiHitGeneratorFromChi2.cc]: Shouldn't this line and lines 141 and 175 be LogDebug() rather than cout?

Comment (makortel) [inline RecoTracker/TkSeedGenerator/plugins/MultiHitGeneratorFromChi2.cc]: Probably yes (also elsewhere in this file where the flag+cout pattern is used), but I'd leave that for another PR if possible.

**Comment by cvuosalo** [comment]
Performance testing in progress...

File: RecoTracker/TkSeedGenerator/plugins/MultiHitGeneratorFromChi2.cc (hunk 1 above)
Comment (cvuosalo) [inline RecoTracker/TkSeedGenerator/plugins/MultiHitGeneratorFromChi2.cc]: @makortel: OK. Please clean up the cout statements in a later PR.

**Comment by cvuosalo** [comment]
+1

For #10016 52de51aeb3b89e08d79f7564ca6cd2393753f42c

Simplification of class structure of hit generators. There should be no change in monitored values.

The code changes are satisfactory, and Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10016/60/summary.html) against baseline CMSSW_7_6_X_2015-07-02-1100 show no significant differences, as expected. A test of workflow 25202.0_TTbar_13 with 70 events also shows on significant differences. Performance seems to improve:

 siPixelClustersPreSplitting      110.6 ms/ev -> 90.6 ms/ev

**Comment by makortel** [comment]
I have no idea how these changes could affect (the performance of) `SiPixelClusterProducer`, so the improvement is probably a fluctuation.

**Comment by davidlange6** [comment]
+1

**Comment by slava77** [comment]
On 7/6/15 1:05 AM, Matti Kortelainen wrote:
> I have no idea how these changes could affect (the performance of)
> |SiPixelClusterProducer|, so the improvement is probably a fluctuation.
> 

this is quite likely from the database latency.

Carl, how did the memory look like?
That part is more interesting, considering the restructuring in the
class structure.

**Comment by cvuosalo** [comment]
@slava:
There were small to negligible memory changes in the comparison against baseline CMSSW_7_6_X_2015-07-02-1100 for workflow 25202.0_TTbar_13. Here they are:

Summary for 70 events
Baseline
Max VSIZ 3685.8 on evt 48 ; max RSS 3106.8 on evt 28

PR 10016
Max VSIZ 3685.85 on evt 33 ; max RSS 3099.8 on evt 28

Product size

 or, B       new, B      delta, B   delta, %   deltaJ, %    branch 
   111448 ->      111697        249      0.2   0.01     TrackingRecHitsOwned_generalTracks__RECO.
  2181334 ->     2181555        221             0.0     ALL BRANCHES

**Comment by slava77** [comment]
On 7/7/15 10:40 AM, Carl Vuosalo wrote:
> @slava <https://github.com/slava>:
> There were small to negligible memory changes in the comparison against
> baseline CMSSW_7_6_X_2015-07-02-1100 for workflow 25202.0_TTbar_13. Here
> they are:
> 
> Summary for 70 events
> Baseline
> Max VSIZ 3685.8 on evt 48 ; max RSS 3106.8 on evt 28
> 
> PR 10016
> Max VSIZ 3685.85 on evt 33 ; max RSS 3099.8 on evt 28

Thanks.
So, there may be a 7MB decrease.
As long as these downward "fluctuations" are accumulating, we are good.
